src/compiler/checker.ts

```diff
@@ -25106,8 +25106,8 @@ namespace ts {
         }
 
         function getContextualTypeForThisPropertyAssignment(binaryExpression: BinaryExpression, kind: AssignmentDeclarationKind): Type | undefined {
-            if (!binaryExpression.symbol) return getTypeOfExpression(binaryExpression.left);
-            if (binaryExpression.symbol.valueDeclaration) {
+            if (!binaryExpression.symbol && kind !== AssignmentDeclarationKind.ModuleExports) return getTypeOfExpression(binaryExpression.left);

```

Comment:
Why can `kind` ever be `ModuleExports` in a function named getContextualTypeFor**ThisProperty**Assignment? Maybe the caller of this function needs the fix instead.

Also I can't figure out which test caused this change.

Reply:
This was causing `module.exports = { ... }` to be contextually typed by its own initializer. I think the function probably just needs to be renamed to `getContextualTypeForSpecialPropertyAssignment`, but I’ll take another look.

Reply:
Oof, this bug actually applies to [every kind of special property assignment](https://www.typescriptlang.org/play?useJavaScript=true#code/LYewJgrgNgpgdDAHgBxAJwC4GcAEBeHAbwCgcycBDAegCoariBfY4pVTLOAMxBHyNLlqdBs2JcIAOwDGGAJYhJOAGIAKAJQDyODAAs5nAEZQKu-iW3bh9QWUYBuJi2Xde525Vo2xL5GhAYAQCeyDDuVl6iznB+AcGhcIjhQpFOxEA).

The fix targeted by this particular change probably does need to move up a level.